File: lottery.go
// lottery.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type Lottery struct {
	numNumbers int
	maxNumber  int
	history    [][]int
	historyFile string
}

func NewLottery(num, max int) *Lottery {
	l := &Lottery{
		numNumbers: num,
		maxNumber:  max,
		historyFile: filepath.Join(os.Getenv("HOME"), ".lottery_history.txt"),
	}
	l.loadHistory()
	return l
}

func (l *Lottery) loadHistory() {
	data, err := os.ReadFile(l.historyFile)
	if err != nil {
		return
	}
	scanner := bufio.NewScanner(strings.NewReader(string(data)))
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		parts := strings.Fields(line)
		var ticket []int
		for _, p := range parts {
			if n, err := strconv.Atoi(p); err == nil {
				ticket = append(ticket, n)
			}
		}
		if len(ticket) > 0 {
			l.history = append(l.history, ticket)
		}
	}
}

func (l *Lottery) saveHistory(ticket []int) {
	l.history = append(l.history, ticket)
	f, err := os.OpenFile(l.historyFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()
	for _, n := range ticket {
		fmt.Fprintf(f, "%d ", n)
	}
	fmt.Fprintln(f)
}

func (l *Lottery) generateTicket() []int {
	rand.Seed(time.Now().UnixNano())
	m := make(map[int]bool)
	for len(m) < l.numNumbers {
		m[rand.Intn(l.maxNumber)+1] = true
	}
	var ticket []int
	for n := range m {
		ticket = append(ticket, n)
	}
	sort.Ints(ticket)
	return ticket
}

func (l *Lottery) checkTicket(ticket, winning []int) (int, []int) {
	var matched []int
	for _, n := range ticket {
		for _, w := range winning {
			if n == w {
				matched = append(matched, n)
				break
			}
		}
	}
	return len(matched), matched
}

func (l *Lottery) showStats() {
	if len(l.history) == 0 {
		fmt.Println(colorize("Нет истории для статистики.", yellow))
		return
	}
	freq := make(map[int]int)
	for _, t := range l.history {
		for _, n := range t {
			freq[n]++
		}
	}
	fmt.Println(colorize("📊 Статистика выпадений:", bold))
	for n := 1; n <= l.maxNumber; n++ {
		if freq[n] > 0 {
			fmt.Printf("  %2d: %d раз\n", n, freq[n])
		}
	}
}

func main() {
	args := os.Args[1:]
	if len(args) == 0 || args[0] == "-h" || args[0] == "--help" {
		fmt.Println("Usage: lottery <num> <max> [-c N] [-w nums] [-s] [-o file] [-v]")
		return
	}
	var num, max, count int
	var winningStr, outputFile string
	statsFlag := false
	verbose := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "-c":
			if i+1 < len(args) {
				count, _ = strconv.Atoi(args[i+1])
				i++
			}
		case "-w":
			if i+1 < len(args) {
				winningStr = args[i+1]
				i++
			}
		case "-s":
			statsFlag = true
		case "-o":
			if i+1 < len(args) {
				outputFile = args[i+1]
				i++
			}
		case "-v":
			verbose = true
		default:
			if num == 0 {
				num, _ = strconv.Atoi(args[i])
			} else if max == 0 {
				max, _ = strconv.Atoi(args[i])
			}
		}
	}
	if num <= 0 || max <= 0 || num > max {
		fmt.Println(colorize("Неверные параметры. Укажите num и max (num <= max).", red))
		return
	}
	if count == 0 {
		count = 1
	}

	game := NewLottery(num, max)

	var winningNumbers []int
	if winningStr != "" {
		parts := strings.Split(winningStr, ",")
		for _, p := range parts {
			n, _ := strconv.Atoi(strings.TrimSpace(p))
			winningNumbers = append(winningNumbers, n)
		}
		if len(winningNumbers) != num {
			fmt.Println(colorize("Количество выигрышных номеров должно совпадать с num.", red))
			return
		}
	}

	if statsFlag {
		game.showStats()
		return
	}

	tickets := make([][]int, count)
	for i := 0; i < count; i++ {
		ticket := game.generateTicket()
		tickets[i] = ticket
		game.saveHistory(ticket)
	}

	outputLines := []string{}
	if verbose {
		for i, ticket := range tickets {
			line := fmt.Sprintf("Билет %d: ", i+1)
			if winningNumbers != nil {
				matches, matched := game.checkTicket(ticket, winningNumbers)
				colored := []string{}
				for _, n := range ticket {
					found := false
					for _, m := range matched {
						if n == m {
							found = true
							break
						}
					}
					if found {
						colored = append(colored, colorize(strconv.Itoa(n), green))
					} else {
						colored = append(colored, strconv.Itoa(n))
					}
				}
				line += strings.Join(colored, " ") + fmt.Sprintf(" (совпадений: %d)", matches)
			} else {
				line += strings.Join(strings.Fields(fmt.Sprint(ticket)), " ")
			}
			outputLines = append(outputLines, line)
		}
	} else {
		for _, ticket := range tickets {
			outputLines = append(outputLines, strings.Join(strings.Fields(fmt.Sprint(ticket)), " "))
		}
	}

	output := strings.Join(outputLines, "\n")
	if outputFile != "" {
		err := os.WriteFile(outputFile, []byte(output), 0644)
		if err == nil {
			fmt.Println(colorize("Результат сохранён в "+outputFile, green))
		} else {
			fmt.Println(colorize("Ошибка записи файла.", red))
		}
	} else {
		fmt.Println(output)
	}
}

func sort.Ints(a []int) {
	for i := 0; i < len(a); i++ {
		for j := i + 1; j < len(a); j++ {
			if a[i] > a[j] {
				a[i], a[j] = a[j], a[i]
			}
		}
	}
}

File: lottery.cpp
// lottery.cpp
#include <iostream>
#include <vector>
#include <set>
#include <string>
#include <random>
#include <algorithm>
#include <fstream>
#include <sstream>
#include <filesystem>
#include <map>

using namespace std;
namespace fs = std::filesystem;

const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string RED = "\033[91m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";
const string BOLD = "\033[1m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

string getHomeDir() {
    const char* home = getenv("HOME");
    if (!home) home = getenv("USERPROFILE");
    return string(home);
}

vector<int> generateTicket(int num, int maxVal) {
    random_device rd;
    mt19937 gen(rd());
    set<int> nums;
    while (nums.size() < num) {
        nums.insert(gen() % maxVal + 1);
    }
    return vector<int>(nums.begin(), nums.end());
}

pair<int, vector<int>> checkTicket(const vector<int>& ticket, const vector<int>& winning) {
    vector<int> matched;
    for (int n : ticket) {
        if (find(winning.begin(), winning.end(), n) != winning.end()) {
            matched.push_back(n);
        }
    }
    return {matched.size(), matched};
}

void showStats(const vector<vector<int>>& history) {
    map<int, int> freq;
    for (auto& t : history) {
        for (int n : t) freq[n]++;
    }
    cout << colorize("📊 Статистика выпадений:", BOLD) << endl;
    for (auto& p : freq) {
        cout << "  " << p.first << ": " << p.second << " раз" << endl;
    }
}

int main(int argc, char* argv[]) {
    int numNumbers = 0, maxNumber = 0;
    int count = 1;
    string winningStr, outputFile;
    bool showStatsFlag = false, verbose = false;

    for (int i=1; i<argc; ++i) {
        string arg = argv[i];
        if (arg == "-h" || arg == "--help") {
            cout << "Usage: lottery <num> <max> [-c N] [-w nums] [-s] [-o file] [-v]" << endl;
            return 0;
        } else if (arg == "-c" && i+1 < argc) {
            count = stoi(argv[++i]);
        } else if (arg == "-w" && i+1 < argc) {
            winningStr = argv[++i];
        } else if (arg == "-s") {
            showStatsFlag = true;
        } else if (arg == "-o" && i+1 < argc) {
            outputFile = argv[++i];
        } else if (arg == "-v") {
            verbose = true;
        } else if (numNumbers == 0) {
            numNumbers = stoi(arg);
        } else if (maxNumber == 0) {
            maxNumber = stoi(arg);
        }
    }
    if (numNumbers <= 0 || maxNumber <= 0 || numNumbers > maxNumber) {
        cerr << colorize("Неверные параметры. Укажите num и max (num <= max).", RED) << endl;
        return 1;
    }

    vector<int> winningNumbers;
    if (!winningStr.empty()) {
        stringstream ss(winningStr);
        string token;
        while (getline(ss, token, ',')) {
            winningNumbers.push_back(stoi(token));
        }
        if ((int)winningNumbers.size() != numNumbers) {
            cerr << colorize("Количество выигрышных номеров должно совпадать с num.", RED) << endl;
            return 1;
        }
    }

    // Загружаем историю
    string histFile = getHomeDir() + "/.lottery_history.txt";
    vector<vector<int>> history;
    if (fs::exists(histFile)) {
        ifstream f(histFile);
        string line;
        while (getline(f, line)) {
            vector<int> ticket;
            stringstream ss(line);
            int n;
            while (ss >> n) ticket.push_back(n);
            if (!ticket.empty()) history.push_back(ticket);
        }
    }

    if (showStatsFlag) {
        showStats(history);
        return 0;
    }

    vector<vector<int>> tickets;
    for (int i=0; i<count; ++i) {
        auto ticket = generateTicket(numNumbers, maxNumber);
        tickets.push_back(ticket);
        history.push_back(ticket);
    }

    // Сохраняем историю
    ofstream f(histFile);
    for (auto& t : history) {
        for (int n : t) f << n << " ";
        f << "\n";
    }

    // Вывод
    vector<string> lines;
    if (verbose) {
        for (size_t i=0; i<tickets.size(); ++i) {
            string line = "Билет " + to_string(i+1) + ": ";
            if (!winningNumbers.empty()) {
                auto res = checkTicket(tickets[i], winningNumbers);
                string colored;
                for (int n : tickets[i]) {
                    if (find(res.second.begin(), res.second.end(), n) != res.second.end()) {
                        colored += colorize(to_string(n), GREEN) + " ";
                    } else {
                        colored += to_string(n) + " ";
                    }
                }
                line += colored + " (совпадений: " + to_string(res.first) + ")";
            } else {
                for (int n : tickets[i]) line += to_string(n) + " ";
            }
            lines.push_back(line);
        }
    } else {
        for (auto& t : tickets) {
            string line;
            for (int n : t) line += to_string(n) + " ";
            lines.push_back(line);
        }
    }

    string output;
    for (auto& l : lines) output += l + "\n";

    if (!outputFile.empty()) {
        ofstream fout(outputFile);
        if (fout) {
            fout << output;
            cout << colorize("Результат сохранён в " + outputFile, GREEN) << endl;
        } else {
            cerr << colorize("Ошибка записи файла.", RED) << endl;
        }
    } else {
        cout << output;
    }
    return 0;
}
